src/stitch.rs

```diff
@@ -79,31 +81,45 @@ impl Iterator for IterStitchedIndexHunks {
         loop {
             // If we're already reading an index, and it has more content, return that.
             if let Some(index_hunks) = &mut self.index_hunks {
+                // An index iterator must be assigned to a band.
+                debug_assert!(self.band_id.is_some());
+
                 for hunk in index_hunks {
                     if let Some(last_entry) = hunk.last() {
                         self.last_apath = Some(last_entry.apath().clone());
                         return Some(hunk);
                     } // otherwise, empty, try the next
                 }
-                if self.archive.band_is_closed(&self.band_id).unwrap_or(false) {
-                    return None;
-                }
+                // There are no more index hunks in the current band.
                 self.index_hunks = None;
-                if let Some(band_id) = previous_existing_band(&self.archive, &self.band_id) {
-                    self.band_id = band_id;
-                } else {
+
+                let band_id = self.band_id.take().expect("last band id should be present");
+                if self.archive.band_is_closed(&band_id).unwrap_or(false) {
+                    // We reached the end of a complete index in this band, 
+                    // so there's no need to look at any earlier bands, and we're done iterating.
                     return None;
                 }
+                
+                // self.band_id might be None afterwards, if there is no previous band.
+                // If so, we're done.
+                self.band_id = previous_existing_band(&self.archive, &band_id);
             }
-            // Start reading this new index and skip forward until after last_apath
-            let mut iter_hunks = Band::open(&self.archive, &self.band_id)
-                .expect("Failed to open band")
-                .index()
-                .iter_hunks();
-            if let Some(last) = &self.last_apath {
-                iter_hunks = iter_hunks.advance_to_after(last)
+
+            if let Some(band_id) = &self.band_id {
+                // Start reading this new index and skip forward until after last_apath
+                let mut iter_hunks = Band::open(&self.archive, &band_id)
+                    .expect("Failed to open band")
+                    .index()
+                    .iter_hunks();
+
+                if let Some(last) = &self.last_apath {
+                    iter_hunks = iter_hunks.advance_to_after(last)
+                }
+                self.index_hunks = Some(iter_hunks);
```

Comment:
Maybe we should 

    debug_assert!(self.index_hunks.is_none(), "existing index hunks iterator would be lost");

Reply:
I'm not sure what what benefit this brings.  
You may be referencing the wrong lines tough.

Reply:
Essentially the idea was to assert that we never ignore index hunks by moving to a different band. I think this is structurally true in this code, but it could have caught future bugs to assert it.